---
title: "Read Data for NCBI GEO Series GSE112679"
author: "Francois.collin@gmail.com"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    code_folding: show
    toc: true
    # does this have an effect
    fig_caption: yes
    # this has no effect
    number_sections: yes
    css: ../docs/_pandocFiles/github-markdown.css
    #css: ../docs/_pandocFiles/pandoc3.css
bibliography: ../docs/_bibFiles/bibFile.bib
csl: ../docs/_bibFiles/cell-numeric.csl
#biblio-style: acm
link-citations: true
vignette: >
 %\VignetteEncoding{UTF-8}
 %\VignetteEngine{knitr::rmarkdown}
---

<!-- TO DO - Styling

formatR::tidy_dir("R")
# or
lintr::lint_package()

save to preTidy first to compare

-->



```{r GlobalOptions, results="hide", include=FALSE, cache=FALSE}
knitr::opts_knit$set(stop_on_error = 2L) #really make it stop
#knitr::dep_auto()
```
<!-- ######################################################################## -->


```{r Prelims, include=FALSE, echo=FALSE, results='hide', message=FALSE,cache=F}

 FN <- "getGSEData"
if(sum(grepl(FN, list.files()))==0) stop("Check FN")

 PREFIX <- ""

 suppressPackageStartupMessages(require(methods))
 suppressPackageStartupMessages(require(rmarkdown))
 suppressPackageStartupMessages(require(bookdown))

 suppressPackageStartupMessages(require(knitr))
 options(stringsAsFactors=F)

 suppressPackageStartupMessages(require(data.table)) 
 options(datatable.fread.datatable=F)

 suppressPackageStartupMessages(require(plyr))
 suppressPackageStartupMessages(require(dplyr))
 suppressPackageStartupMessages(require(magrittr))

 # Shotcuts for knitting and redering while in R session (Invoke interactive R from R/Scripts folder)
 kk <- function(n='') knitr::knit2html(paste("t", n, sep=''), envir=globalenv(),
       output=paste(FN,".html", sep=''))

 rr <- function(n='') rmarkdown::render(paste("t", n, sep=''), envir=globalenv(),
       output_file=paste(FN,".html", sep='')) ##, output_dir='Scripts')

 bb <- function(n='') browseURL(paste(FN,".html", sep=''))

 # The usual shotcuts
 zz <- function(n='') source(paste("t", n, sep=''))

 # Using relative paths:
 # Assuming script is run from GSE112679/data-raw/
 WRKDIR <- ('..')

 # Not needed if path is relative ...
 if(!file.exists(WRKDIR)) stop("WRKDIR ERROR: ", WRKDIR)

 # do once
 #setwd(WRKDIR)

 # file rmarkdown file management options: cache, figures
 cache_DIR <- file.path('cache/rdGEData/')
 suppressPackageStartupMessages(dir.create(cache_DIR, recursive=T))
 opts_chunk$set(cache.path=cache_DIR)

 figure_DIR <- file.path('figures/rdGEData/')
 suppressPackageStartupMessages(dir.create(figure_DIR, recursive=T))
 opts_chunk$set(fig.path=paste0(figure_DIR, PREFIX))

 data_DIR <- '../data'
 if(!file.exists(data_DIR)) stop("data_DIR ERROR: ", data_DIR)


```
<!-- ######################################################################## -->


*** 

# Abstract

This script assembles R object from data downloaded from 
[NCBI/Gene Expression Omnibus Database (GSE112679)](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112679).
The R objects will be assembled into an R data package for
easy access.  The objects are constructed to parallel the
analyses of these data as reported in
Cai et al. (2019) [@Cai:2019aa]).  



# Set Analysis Parameters
 
```{r setAnalysisParameters, echo=T, cache=F}

 GSE_ID <- 'GSE112679'

 # extData
 ################################
 if(file.exists(file.path('../../extData'))) 
 EXT_DIR <- file.path('../../extData') else stop("Speficy EXT_DIR")

 # GSE_Data
 ###################################
 if(file.exists(file.path(EXT_DIR, GSE_ID)))
 GSE_DATA_DIR <- file.path(EXT_DIR,GSE_ID) else stop("Speficy GSE_DATA_DIR")

 # SampleDesc
 ###################################
 if(file.exists(file.path(GSE_DATA_DIR, paste0(GSE_ID,'_series_matrix.txt'))))
 SAMPLE_DESC_FILE <- file.path(GSE_DATA_DIR, paste0(GSE_ID,'_series_matrix.txt')) else 
 stop("Speficy SAMPLE_DESC_FILE")
 
```

Define some utility functions.  

```{r utilFunctions, cache=F}
 # The sequencing files have two formats
 ## _[TU,TI]_ files have columns Geneid, Chr, Start, End, Strand, Length, Count
 ## _seq_ fales have columns Geneid, count
 
#############################################
#' Function to read the *wide* format
#'
#' @param seqFile A character string specifying a file location
#' @param Verbose A logical flag controlling printing
readSeqWfmt <- function(seqFile, Verbose=F){
 seqFile.frm <- fread(seqFile)
 names(seqFile.frm) <- c('GeneId','Chr', 'Start', 'End', 'Strand', 'Length', 'Count')

 # Reformat for ordering
 seqFile.frm$Chr <- 
  as.character(factor(seqFile.frm$Chr,
               levels=paste0('chr', c(1:22,'X','Y','M')),
               labels=paste0('chr', c(formatC(1:22,width=2,flag='0'),'X','Y','M'))))

 rownames(seqFile.frm) <-
 with(seqFile.frm, 
       paste(Chr, Strand, Start, End, GeneId, sep='~'))

 seqFile.frm <- seqFile.frm[,-(1:6),drop=F]
 if(Verbose) print(seqFile.frm[1:6,,drop=F])
 if(Verbose) print(dim(seqFile.frm))

 seqFile.frm}

#############################################
#' Function to read the *narrow* format
#'
#' @param seqFile A character string specifying a file location
#' @param Verbose A logical flag controlling printing
readSeqNfmt <- function(seqFile, Verbose=F){
 seqFile.frm <- fread(seqFile)
 names(seqFile.frm) <- c('GeneId','Count')

 rownames(seqFile.frm) <- seqFile.frm$GeneId

 seqFile.frm <- seqFile.frm[,-1,drop=F]
 if(Verbose) print(seqFile.frm[1:6,,drop=F])
 if(Verbose) print(dim(seqFile.frm))

 seqFile.frm}

#############################################
#' Function to assemble a table of counts 
#' from a named vector of files and a reading function
#'
#' @param countFiles A character string vector of file locations
#' @param Verbose A logical flag controlling printing
#' @param readSeqFun A fucntion to use to read the countFiles
getCounts <- function(countFiles, Verbose=F, readSeqFun){
  
  # Read first seqFile
  featureCount <- readSeqFun(countFiles[1], Verbose=Verbose)

  # Read the rest /* appening columns in a loop is not efficient ... */
  for(FF in countFiles[-1])
  featureCount <- cbind(featureCount, 
  readSeqFun(FF, Verbose=Verbose)[rownames(featureCount),])

  colnames(featureCount) <- names(countFiles)
  as.matrix(featureCount)}

#############################################
#' Get sequencing data for sampDesc$sampType in sampTypeSel
#'
#' @param selVar A character string specifying the selction variable
#' @param selVal A character string vector specifying selected sampType values
#' @param readSeqFun A function to use to read the selected countFiles
getData <- function(selVar, selVal, readSeqFun) {
 # Should validate selVar, selVal
 selected_sampDesc <- sampDesc[sampDesc[,selVar] %in% selVal,] 
 rownames(selected_sampDesc) <- selected_sampDesc$geoAcc

 selected_countFiles <- selected_sampDesc$fileName
 names(selected_countFiles) <- sapply(strsplit(selected_countFiles, split='_'), '[',1)

 selected_countFiles_ndx <- match(selected_countFiles, GSE_genebodyCountFiles)

 if(sum(is.na(selected_countFiles_ndx))) {
   cat("Some count files are missing:\n")
   prnt(selected_countFiles[is.na(selected_countFiles_ndx)])
 }

 selected_countFiles <- intersect(selected_countFiles, GSE_genebodyCountFiles)
 names(selected_countFiles) <- sapply(strsplit(selected_countFiles, split='_'), '[',1)

 cat("\nFound", length(selected_countFiles), 'count files\n')
 #print(selected_countFiles)

 selected_countFiles_loc <- file.path(GSE_DATA_DIR, selected_countFiles)
 names(selected_countFiles_loc) <- names(selected_countFiles) 
 selected_featureCount <- getCounts(selected_countFiles_loc, Verbose=F, readSeqFun)
 selected_featureCount}

```

## Get Sample Description Data

```{r readSampDesc,cache=T, cach.vars='sampDesc', eval=T, echo=T, fig.height=6, fig.width=11}

 suppressPackageStartupMessages(require(GEOquery))

 GSEMatrix_obj <- getGEO(GSE_ID, GSEMatrix=T)
 show(GSEMatrix_obj)

 #DT::datatable(pData(phenoData(GSEMatrix_obj[[1]])))

 KEY_FIELDS <- grep(':ch1$', names(pData(phenoData(GSEMatrix_obj[[1]]))), value=T)

 sampType_vec <- sapply(
 strsplit(as.character(pData(phenoData(GSEMatrix_obj[[1]]))[,"title"]), 
 split='_'), function(X) X[1])

 sampNo_vec <- sapply(
 strsplit(as.character(pData(phenoData(GSEMatrix_obj[[1]]))[,"title"]), 
 split='_'), function(X) X[2])

 bioSamples_vec <- sapply(
 strsplit(as.character(pData(phenoData(GSEMatrix_obj[[1]]))[,"relation"]), 
 split='/'), function(X) rev(X)[1])

 SRA_vec <- sapply(
 strsplit(as.character(pData(phenoData(GSEMatrix_obj[[1]]))[,"relation.1"]), 
 split='='), function(X) rev(X)[1])

 fileName_vec <- sapply(
 strsplit(as.character(pData(phenoData(GSEMatrix_obj[[1]]))[,"supplementary_file_1"]), 
 split='/'), function(X) rev(X)[1])

 sampDesc <- data.frame(
   geoAcc=pData(phenoData(GSEMatrix_obj[[1]]))[,"geo_accession"],
   title=pData(phenoData(GSEMatrix_obj[[1]]))[,"title"],
   sampType=sampType_vec,
   sampNo=sampNo_vec,
   bioSample=bioSamples_vec, 
   SRA=SRA_vec,
   fileName=fileName_vec,
   pData(phenoData(GSEMatrix_obj[[1]]))[, KEY_FIELDS])

 names(sampDesc) <- 
 sub('bclc.stage.ch1', 'stage',
 sub('diagnosis.ch1', 'Dx',
 sub('tissue.subtype.ch1', 'tissueSubtype',
 sub('tissue.ch1', 'tissue',
 sub('training.validation.group.ch1', 'trainValGroup',
 names(sampDesc))))))

 # Make sure fileName and geoAcc match
 fileNamegeoAcc.vec <- sapply(
 strsplit(as.character(sampDesc$fileName), split='_'), '[',1)
 if(sum(fileNamegeoAcc.vec!= sampDesc$geoAcc))
 stop("geoAcc/fileName Mismatch")

 str(sampDesc)
 
 cat("geoAcc is unique - use as rownames:\n")
 with(sampDesc, table(table(geoAcc, exclude=NULL)))
 rownames(sampDesc) <- sampDesc$geoAcc
 
 cat("SRA is unique:\n")
 with(sampDesc, table(table(SRA, exclude=NULL)))
 
 cat("bioSample is unique:\n")
 with(sampDesc, table(table(bioSample, exclude=NULL)))
 
 cat("title is unique:\n")
 with(sampDesc, table(table(title, exclude=NULL)))

 cat("Some Samples Match bu sampNo:")
 with(sampDesc, table(table(sampNo, exclude=NULL)))

 # NOTE: examination of the data indicate that sampNo cannot be used
 # to match Blood with TU or TI samples

DEPRICATED <- function() {
 NoSamp.tbl <- with(sampDesc, table(sampNo, exclude=NULL))

 sampDesc <- merge(
  data.frame(sampNo=names(nSamp.tbl), nSamp=as.vector(nSamp.tbl)),
  sampDesc, by='sampNo', all.y=T)
}#DEPRICATED

 sampDesc <- sampDesc[with(sampDesc,
  order(as.numeric(sampNo), title)),]


 # Shorten trainValGroup
 sampDesc$trainValGroup <- 
 sub('Training', 'Train',
 sub('Validation', 'Val',
 sampDesc$trainValGroup))

 # Shorten Dx
 sampDesc$Dx <- 
 sub('Benign liver lesions', 'Benign',
 sub('Liver cirrhosis', 'Cirrhosis',
 sampDesc$Dx))


 trainValGroupDX.tbl <-  
 with(sampDesc, 
 table(Dx_tissue=paste0(Dx,'_',tissue), trainValGroup, exclude=NULL))

 trainValGroupDX.tbl

 # For consistency with previous code, we will use Outcome as an alias to Dx,
 # and sampID as an alias to geoAcc
 sampDesc$outcome <- sampDesc$Dx
 sampDesc$sampID <- sampDesc$geoAcc

 cat("Cai et al. combine Beign+Healthy and Cirrhosis+HCC\n")
 cat("Create secondary outcome\n")
 sampDesc$outcome2 <- with(sampDesc, 
 ifelse(Dx %in% c("Benign", "Healthy"), 'BenignHealthy',
 ifelse(Dx %in% c("Cirrhosis", "CHB"), 'CirrhosisCHB', Dx)))

 with(sampDesc, table(trainValGroup, outcome2, exclude=NULL))
 
 cat("Also want outcome3 == HCC or nonHCC\n")
 sampDesc$outcome3 <- with(sampDesc,
 ifelse(Dx == 'HCC', 'HCC', 'nonHCC'))

 with(sampDesc, table(trainValGroup, outcome3, exclude=NULL))


```

```{r DTsampDesc}

 DT::datatable(sampDesc,  options=list(pageLength = 18))

```


## Get Count Data


NOTE: examination of the data indicate that `sampNo` cannot be used
to match Blood with TU or TI samples (Dx does not agree!).  It is also unclear whether 
`sampNo` can be used to match TU with TI.  Examination of the 
5hmC gene body profile similarities should tell us.  

``` Explore this in vignette ```

There are too many samples in this dataset to store into a single count object.
We will create 4 count matrices, TUTI, Train, Val_1, Val_2, 
corresponding to the matched tumor-tissue samples, the plasma train set and
the two plasma validation sets.  ***Note that there will be some overlap 
between TUTI and Train and Val_1***.

<!-- IDEA
We will instead read each vector and store to the file system as a separate 
object.  Matrices will be assmbled as needed.  
-->


### Read sequencing data for `TUTI` set, the matched tummor, tissue samples.

```{r  readTUTI, eval=T, cache=T, cache.vars=c('TUTI_featureCount', 'genes_annot'),message=F}
 ####,results='asis'}
 suppressPackageStartupMessages(require(edgeR))

 GSE_genebodyCountFiles <- list.files(GSE_DATA_DIR, 'genebody')

 TUTI_featureCount <- getData(selVar='sampType', selVal=c('TU', 'TI'), readSeqFun=readSeqWfmt)

 cat("dim(TUTI_featureCount):\n")
 print(dim(TUTI_featureCount))
 print(TUTI_featureCount[1:5, 1:5])

 # Extract genes_annot from rownames of TUTI_countFiles
 genes_annot <-
 data.frame(Symbol=sapply(strsplit(rownames(TUTI_featureCount), split='~'), function(x) rev(x)[1]),
             Chr=sapply(strsplit(rownames(TUTI_featureCount), split='~'), '[', 1),
             Strand=sapply(strsplit(rownames(TUTI_featureCount), split='~'), '[', 2),
             Start=as.numeric(sapply(strsplit(rownames(TUTI_featureCount), split='~'), '[', 3)),
             Stop=as.numeric(sapply(strsplit(rownames(TUTI_featureCount), split='~'), '[', 4)),
             Length=sapply(strsplit(rownames(TUTI_featureCount), split='~'),
       function(x) as.numeric(x[4]) - as.numeric(x[3]) +1))
 rownames(genes_annot) <- rownames(TUTI_featureCount)
 
```

### Add GC content to genes_annot

```{r addGC, eval=T, cache=T, cache.vars=c('genes_annot'),message=F}
 library(BSgenome.Hsapiens.UCSC.hg19)
 library(TxDb.Hsapiens.UCSC.hg19.knownGene)

 genes_gr <- GRanges(
 seqnames=sub('chr0','chr',genes_annot$Chr),
 IRanges(start=genes_annot$Start, end=genes_annot$Stop),
 strand=genes_annot$Strand)

 genes_seq <- Biostrings::getSeq(BSgenome.Hsapiens.UCSC.hg19, genes_gr)
 genes_alf <- Biostrings::alphabetFrequency(genes_seq, collapse=F)[, c("A", "C", "G", "T")]

 genes_annot <- data.frame(genes_annot, genes_alf)

```



<br/>

### Read sequencing data for samples in `Train` set

```{r  readTrain, eval=T, cache=T, cache.vars='Train_featureCount',message=F}
 ####,results='asis'}
 suppressPackageStartupMessages(require(edgeR))

 GSE_genebodyCountFiles <- list.files(GSE_DATA_DIR, 'genebody')

 Train_featureCount <- getData(selVar='trainValGroup', selVal='Train', readSeqFun=readSeqNfmt)

 cat("dim(Train_featureCount):\n")
 print(dim(Train_featureCount))
 print(Train_featureCount[1:5, 1:5])

 # Verify rows match genes_annot
 genes_ndx <- match(rownames(Train_featureCount), genes_annot$Symbol)
 if(sum(is.na(genes_ndx))) stop("Mismatched Genes")
 if(sum(duplicated(genes_ndx))) stop("Duplicated Genes")

```

 
<br/>

### Read sequencing data for samples in `Val-1` set


```{r  readVal-1, eval=T, cache=T, cache.vars='Val1_featureCount',message=F}
 ####,results='asis'}
 suppressPackageStartupMessages(require(edgeR))

 GSE_genebodyCountFiles <- list.files(GSE_DATA_DIR, 'genebody')

 Val1_featureCount <- getData(selVar='trainValGroup', selVal='Val-1', readSeqFun=readSeqNfmt)

 cat("dim(Val1_featureCount):\n")
 print(dim(Val1_featureCount))
 print(Val1_featureCount[1:5, 1:5])

 # Verify rows match genes_annot
 genes_ndx <- match(rownames(Val1_featureCount), genes_annot$Symbol)
 if(sum(is.na(genes_ndx))) stop("Mismatched Genes")
 if(sum(duplicated(genes_ndx))) stop("Duplicated Genes")

```


### Read sequencing data for samples in `Val-2` set


```{r  readVal-2, eval=T, cache=T, cache.vars='Val2_featureCount',message=F}
 ####,results='asis'}
 suppressPackageStartupMessages(require(edgeR))

 GSE_genebodyCountFiles <- list.files(GSE_DATA_DIR, 'genebody')

 Val2_featureCount <- getData(selVar='trainValGroup', selVal='Val-2', readSeqFun=readSeqNfmt)

 cat("dim(Val2_featureCount):\n")
 print(dim(Val2_featureCount))
 print(Val2_featureCount[1:5, 1:5])

 # Verify rows match genes_annot
 genes_ndx <- match(rownames(Val2_featureCount), genes_annot$Symbol)
 if(sum(is.na(genes_ndx))) stop("Mismatched Genes")
 if(sum(duplicated(genes_ndx))) stop("Duplicated Genes")

```

## Save all to `r GSE_ID`


```{r save, cache=T, cache.vars=''}
 # THIS DOESNT WORK
 #usethis::use_data(get(paste0(GSE_ID, '_sampDesc')), overwrite=T)

 # DO WE HAVE TO DO THIS?
 #usethis::use_data(GSE112679_sampDesc, overwrite=T)

THIS.DONT.WORK <- function() {
 # put all data - sampDesc & counts - into one object!
 save(list=c('sampDesc', 'TUTI_featureCount', 'Train_featureCount',
             'Val1_featureCount', 'Val2_featureCount', 'genes_annot'),
      file=file.path(data_DIR, paste0(GSE_ID, '.rda')))
 }


# save sampDesc
assign(paste0(GSE_ID,'_sampDesc'), sampDesc)
save(list=paste0(GSE_ID,'_sampDesc'),
     file=file.path(data_DIR, paste0(GSE_ID,'_sampDesc.rda')),
  compress='xz')


# save counts
assign(paste0(GSE_ID,'_TUTI_featureCount'), TUTI_featureCount)
save(list=paste0(GSE_ID,'_TUTI_featureCount'),
     file=file.path(data_DIR, paste0(GSE_ID,'_TUTI_featureCount.rda')),
  compress='xz')

assign(paste0(GSE_ID,'_Train_featureCount'), Train_featureCount)
save(list=paste0(GSE_ID,'_Train_featureCount'),
     file=file.path(data_DIR, paste0(GSE_ID,'_Train_featureCount.rda')),
  compress='xz')

assign(paste0(GSE_ID,'_Val1_featureCount'), Val1_featureCount)
save(list=paste0(GSE_ID,'_Val1_featureCount'),
     file=file.path(data_DIR, paste0(GSE_ID,'_Val1_featureCount.rda')),
  compress='xz')

assign(paste0(GSE_ID,'_Val2_featureCount'), Val2_featureCount)
save(list=paste0(GSE_ID,'_Val2_featureCount'),
     file=file.path(data_DIR, paste0(GSE_ID,'_Val2_featureCount.rda')),
  compress='xz')

# save genes_annot
assign(paste0(GSE_ID,'_genes_annot'), genes_annot)
save(list=paste0(GSE_ID,'_genes_annot'),
     file=file.path(data_DIR, paste0(GSE_ID,'_genes_annot.rda')),
  compress='xz')


```
 

# References
<div id="refs"></div>



```{r, echo=FALSE}
 sessionInfo()
```


```{r, echo=FALSE}
  knit_exit()
```

### ARCHIVAL CODE BELOW
<!-- ######################################################################## -->


<!-- To run
# nohup Rscript -e "knitr::knit2html('getGSEData.Rmd')" > getGSEData.log  &

# Or
# nohup Rscript -e "rmarkdown::render('getGSEData.Rmd')" > getGSEData.log  &

-->

